master/static/srv/update_aggregated_allocation.sql

```diff
@@ -5,67 +5,44 @@ WITH const AS (
             ($1 :: timestamptz + interval '1 day')
         ) AS period
 ),
--- Workloads that had any overlap with the target interval, along with the length of the overlap of
+-- Allocations that had any overlap with the target interval, along with the length of the overlap of
 -- their time with the requested period.
-workloads AS (
+allocs_in_range AS (
     SELECT
-        all_workloads.trial_id,
+        a.*,
         extract(
             epoch
             FROM
                 -- `*` computes the intersection of the two ranges.
-                upper(const.period * range) - lower(const.period * range)
-        ) * (
-            experiments.config -> 'resources' ->> 'slots_per_trial'
-        ) :: float AS seconds
+                upper(const.period * a.range) - lower(const.period * a.range)
+        ) * a.slots :: float AS seconds
     FROM
         (
             SELECT
-                trial_id,
+                *,
                 tstzrange(start_time, end_time) AS range
             FROM
-                steps
-            UNION ALL
-            SELECT
-                trial_id,
-                tstzrange(start_time, end_time) AS range
-            FROM
-                validations
-            UNION ALL
-            SELECT
-                trial_id,
-                tstzrange(start_time, end_time) AS range
-            FROM
-                checkpoints
-        ) AS all_workloads,
-        trials,
-        experiments,
+                allocations
+        ) AS a,
         const
     WHERE
         -- `&&` determines whether the ranges overlap.
-        const.period && all_workloads.range
-        AND all_workloads.trial_id = trials.id
-        AND trials.experiment_id = experiments.id
-        AND (
-            -- Sometimes experiments end uncleanly, leaving workloads permanently without end
-            -- times. Anything in that state must be excluded or it'll show up in all the statistics
-            -- forevermore.
-            upper(all_workloads.range) IS NOT NULL
-            OR experiments.state = 'ACTIVE'
-        )
+        const.period && a.range
```

Comment:
Alright, if there's no way now to know whether something is still active or not, this is good. Would such a thing be feasible to add at some point?

Reply:
I mean, the way I see it, if we can't guarantee to write the end time, we can't guarantee to write some state change to 'final' that means the end time should be assumed invalid and it should not be aggregated (e.g. like the state `experiment.state = 'ERRORED' AND upper(all_workloads.range) IS NULL` previously would've been caused by a trial actor crash and excluded, it's entirely possible in the event of database network issues that the trial _and_ experiment actor could've crashed and left us as `experiment.state = 'ACTIVE' AND upper(all_workloads.range) IS NULL` and the workloads would've counted forever right?

I think the best we can do, probably, is to mark all active jobs/tasks/allocations end_time = current_timestamp on a master restart, as a "just in case" sort of measure.

Reply:
To get the old behavior (2 db writes need to fail to cause issues), we could take greatest(upper(allocation_range), upper(task_range)). I just assumed db write failures would be highly correlated in such a short time span so that complexity wouldn't really be worth the hassle.

Reply:
Yeah, clearing out stuff on restart sounds good enough to me and is roughly what I was imagining. Can't do anything until the next restart anyway. I think that matches well enough with what happens with an experiment, where if it stays in ACTIVE because of a master crash, then the next time the master starts, it'll either crash and move out of ACTIVE soon or continue running and eventually finish.

Reply:
I'll go ahead and just tack that onto this pr, since it's just running `update allocations set end_time = current_timestamp where end_time is null;`